presto-main/src/main/java/io/prestosql/sql/planner/optimizations/PropertyDerivations.java

```diff
@@ -438,11 +440,59 @@ public ActualProperties visitJoin(JoinNode node, List<ActualProperties> inputPro
-                    // a hash-partitioned join can produce nulls in case of a lack of matches
-                    return ActualProperties.builder()
-                            .global(probeProperties.isSingleNode() ? singleStreamPartition() : arbitraryPartition())
-                            .build();
+                    if (SystemSessionProperties.isDeriveFullJoinProperties(session)) {
+                        // the join keys might not be output by the join node, so we need to see if we can express the partitioning scheme
+                        // for the join in terms of its outputs
+                        probeProperties = probeProperties.translate(column -> filterIfMissing(node.getOutputSymbols(), column));
+                        buildProperties = buildProperties.translate(column -> filterIfMissing(node.getOutputSymbols(), column));
+
+                        if (!probeProperties.getNodePartitioning().isPresent() || !buildProperties.getNodePartitioning().isPresent()) {
+                            return ActualProperties.builder()
+                                    .global(probeProperties.isSingleNode() ? singleStreamPartition() : arbitraryPartition())
+                                    .build();
+                        }
+
+                        List<Partitioning.ArgumentBinding> newBindings = new ArrayList<>();
+                        for (int i = 0; i < probeProperties.getNodePartitioning().get().getArguments().size(); i++) {
+                            Partitioning.ArgumentBinding left = probeProperties.getNodePartitioning().get().getArguments().get(i);
+                            Partitioning.ArgumentBinding right = buildProperties.getNodePartitioning().get().getArguments().get(i);
+
+                            if (left.isVariable() && right.isVariable()) {
+                                boolean found = false;
+                                for (JoinNode.EquiJoinClause equality : node.getCriteria()) {
+                                    if (equality.getLeft().equals(left.getColumn()) && equality.getRight().equals(right.getColumn()) ||
+                                            equality.getLeft().equals(right.getColumn()) && equality.getRight().equals(left.getColumn())) {
+                                        found = true;
+                                        newBindings.add(Partitioning.ArgumentBinding.expressionBinding(new CoalesceExpression(ImmutableList.of(equality.getLeft().toSymbolReference(), equality.getRight().toSymbolReference()))));
+                                    }
+                                }
+                                checkState(found, "Partitioning inconsistent with join criteria");
+                            }
+                            else if (left.isConstant() && right.isVariable()) {
+                                checkState(buildProperties.getConstants().get(right.getColumn()).equals(left.getConstant()));
```

Comment:
that is odd for me that even though we know `right` is constant, we don't store that info in bindings.

There is a comment in `Partitioning`:
```
            // As a last resort, check for a constant mapping for the symbol
            // Note: this MUST be last because we want to favor the symbol representation
            // as it makes further optimizations possible.
``` 
But isn't that having a constant the best representation?

Reply:
I don't remember the context around that comment, but yeah, knowing things are constants is usually better. It may have to do with the ability to reason about and match partitionings when dealing with joins. Let me take a deeper look.